### PR TITLE
Create `@next/package-manager-utils` to unify package manager utilities across the repo.

### PR DESCRIPTION
**Pull request by Ethan-Arrowood**
I've noticed that we have duplicate, inconsistent, and untested logic for package manager utilities. This PR creates a new package (for internal use only?) that unifies these utilities in a single location. This enables us to make updates to them easier, have consistent usage across the repo, and properly test their functionality. Today we rely on our integration/e2e testing only and they do not cover all cases such as the issue I fixed here: https://github.com/vercel/next.js/pull/64418.

The files in consideration include:
- [packages/next/src/lib/helpers/get-pkg-manager.ts](https://github.com/vercel/next.js/tree/canary/packages/next/src/lib/helpers/get-pkg-manager.ts)
- [packages/next/src/lib/helpers/get-registry.ts](https://github.com/vercel/next.js/tree/canary/packages/next/src/lib/helpers/get-registry.ts)
- [packages/next/src/lib/helpers/get-online.ts](https://github.com/vercel/next.js/tree/canary/packages/next/src/lib/helpers/get-online.ts)
- [packages/next/src/lib/helpers/get-npx-commands.ts](https://github.com/vercel/next.js/tree/canary/packages/next/src/lib/helpers/get-npx-commands.ts)
- [packages/next/src/lib/helpers/install.ts](https://github.com/vercel/next.js/tree/canary/packages/next/src/lib/helpers/install.ts)
- [packages/create-next-app/helpers/get-pkg-manager.ts](https://github.com/vercel/next.js/tree/canary/packages/create-next-app/helpers/get-pkg-manager.ts)
- [packages/create-next-app/helpers/is-online.ts](https://github.com/vercel/next.js/tree/canary/packages/create-next-app/helpers/is-online.ts)
- [packages/create-next-app/helpers/install.ts](https://github.com/vercel/next.js/tree/canary/packages/create-next-app/helpers/install.ts)

The plan is to introduce these changes in parts. First, I'd like to unify the major overlap (`get-pkg-manager`, `is-online`, and `install`). Afterwards, I'll work on moving some of the other utilities over. I also need to look over the remaining packages and see if any of them have similar logic to include in this work.

### CODE REVIEW

**Comment by Ethan-Arrowood** [comment]
* **#65096** <a href="https://app.graphite.dev/github/pr/vercel/next.js/65096?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `canary`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @Ethan-Arrowood and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Comment by ijjk** [comment]
# Stats from current PR

<details open>
<summary><strong>Default Build</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong></summary>

|  | vercel/next.js canary  | vercel/next.js add-next-package-manager-utils | Change |
| - | - | - | - |
| buildDuration | 28.6s | 24.5s | N/A |
| buildDurationCached | 14.7s | 12.1s | N/A |
| nodeModulesSize | 359 MB | 359 MB | ✓ |
| nextStartRea..uration (ms) | 733ms | 647ms | N/A |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js add-next-package-manager-utils | Change |
| - | - | - | - |
| 1103-HASH.js gzip | 31.8 kB | 31.8 kB | N/A |
| 1a9f679d-HASH.js gzip | 53.5 kB | 53.5 kB | N/A |
| 335-HASH.js gzip | 5.09 kB | 5.09 kB | N/A |
| 7953.HASH.js gzip | 169 B | 169 B | ✓ |
| framework-HASH.js gzip | 45.2 kB | 45.2 kB | ✓ |
| main-app-HASH.js gzip | 231 B | 229 B | N/A |
| main-HASH.js gzip | 31.5 kB | 31.5 kB | N/A |
| webpack-HASH.js gzip | 1.65 kB | 1.65 kB | N/A |
| Overall change | 45.4 kB | 45.4 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js add-next-package-manager-utils | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | vercel/next.js add-next-package-manager-utils | Change |
| - | - | - | - |
| _app-HASH.js gzip | 192 B | 193 B | N/A |
| _error-HASH.js gzip | 192 B | 193 B | N/A |
| amp-HASH.js gzip | 507 B | 511 B | N/A |
| css-HASH.js gzip | 341 B | 343 B | N/A |
| dynamic-HASH.js gzip | 2.52 kB | 2.52 kB | ✓ |
| edge-ssr-HASH.js gzip | 266 B | 265 B | N/A |
| head-HASH.js gzip | 362 B | 365 B | N/A |
| hooks-HASH.js gzip | 392 B | 392 B | ✓ |
| image-HASH.js gzip | 4.32 kB | 4.32 kB | N/A |
| index-HASH.js gzip | 268 B | 268 B | ✓ |
| link-HASH.js gzip | 2.69 kB | 2.7 kB | N/A |
| routerDirect..HASH.js gzip | 329 B | 328 B | N/A |
| script-HASH.js gzip | 397 B | 397 B | ✓ |
| withRouter-HASH.js gzip | 324 B | 324 B | ✓ |
| 1afbb74e6ecf..834.css gzip | 106 B | 106 B | ✓ |
| Overall change | 4 kB | 4 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | vercel/next.js add-next-package-manager-utils | Change |
| - | - | - | - |
| _buildManifest.js gzip | 483 B | 484 B | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js add-next-package-manager-utils | Change |
| - | - | - | - |
| index.html gzip | 528 B | 528 B | ✓ |
| link.html gzip | 540 B | 540 B | ✓ |
| withRouter.html gzip | 523 B | 524 B | N/A |
| Overall change | 1.07 kB | 1.07 kB | ✓ |

</details>

<details>
<summary><strong>Edge SSR bundle Size</strong></summary>

|  | vercel/next.js canary  | vercel/next.js add-next-package-manager-utils | Change |
| - | - | - | - |
| edge-ssr.js gzip | 94.7 kB | 94.7 kB | N/A |
| page.js gzip | 3.07 kB | 3.06 kB | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Middleware size</strong></summary>

|  | vercel/next.js canary  | vercel/next.js add-next-package-manager-utils | Change |
| - | - | - | - |
| middleware-b..fest.js gzip | 623 B | 622 B | N/A |
| middleware-r..fest.js gzip | 156 B | 156 B | ✓ |
| middleware.js gzip | 27.9 kB | 27.9 kB | N/A |
| edge-runtime..pack.js gzip | 839 B | 839 B | ✓ |
| Overall change | 995 B | 995 B | ✓ |

</details>

<details>
<summary><strong>Next Runtimes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js add-next-package-manager-utils | Change |
| - | - | - | - |
| app-page-exp...dev.js gzip | 171 kB | 171 kB | ✓ |
| app-page-exp..prod.js gzip | 98.4 kB | 98.4 kB | ✓ |
| app-page-tur..prod.js gzip | 99.9 kB | 99.9 kB | ✓ |
| app-page-tur..prod.js gzip | 94.3 kB | 94.3 kB | ✓ |
| app-page.run...dev.js gzip | 157 kB | 157 kB | ✓ |
| app-page.run..prod.js gzip | 93 kB | 93 kB | ✓ |
| app-route-ex...dev.js gzip | 21.5 kB | 21.5 kB | ✓ |
| app-route-ex..prod.js gzip | 15.2 kB | 15.2 kB | ✓ |
| app-route-tu..prod.js gzip | 15.2 kB | 15.2 kB | ✓ |
| app-route-tu..prod.js gzip | 15 kB | 15 kB | ✓ |
| app-route.ru...dev.js gzip | 21.3 kB | 21.3 kB | ✓ |
| app-route.ru..prod.js gzip | 15 kB | 15 kB | ✓ |
| pages-api-tu..prod.js gzip | 9.55 kB | 9.55 kB | ✓ |
| pages-api.ru...dev.js gzip | 9.82 kB | 9.82 kB | ✓ |
| pages-api.ru..prod.js gzip | 9.55 kB | 9.55 kB | ✓ |
| pages-turbo...prod.js gzip | 21.5 kB | 21.5 kB | ✓ |
| pages.runtim...dev.js gzip | 22.1 kB | 22.1 kB | ✓ |
| pages.runtim..prod.js gzip | 21.5 kB | 21.5 kB | ✓ |
| server.runti..prod.js gzip | 51.6 kB | 51.6 kB | ✓ |
| Overall change | 962 kB | 962 kB | ✓ |

</details>

<details>
<summary><strong>build cache</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | vercel/next.js add-next-package-manager-utils | Change |
| - | - | - | - |
| 0.pack gzip | 1.61 MB | 1.61 MB | N/A |
| index.pack gzip | 112 kB | 112 kB | ⚠️ +244 B |
| Overall change | 112 kB | 112 kB | ⚠️ +244 B |

</details>

<details>
<summary><strong>Diff details</strong></summary>

<details>
<summary>Diff for <strong>edge-ssr.js</strong></summary>

Diff too large to display
</details>

</details>

</details>
Commit: c0b00f4f5ab10ddfb6bf940e532de93eae51a598

File: packages/next-package-manager-utils/index.ts [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+        dev: '--save-dev',
+        prod: '--save-prod',
+        offline: '--offline',
+      },
+    },
+  },
+  {
+    name: 'yarn',
+    lockfile: 'yarn.lock',
+    registry: 'registry.yarnokg.com',
```
Comment (ykzts) [inline packages/next-package-manager-utils/index.ts:62]:
```suggestion
    registry: 'registry.yarnpkg.com',
```

typo?

**Comment by ijjk** [comment]
## Failing test suites 
Commit: edc7b79f00705bf09d4817b54f9723ee55882798

`TURBOPACK=1 pnpm test-start test/e2e/app-dir/app-client-cache/client-cache.test.ts` (turbopack)
- app dir client cache semantics > prefetch={true} > should prefetch again after 5 mins if the link is visible again

<details>
<summary>Expand output</summary>

  ● app dir client cache semantics › prefetch={true} › should prefetch again after 5 mins if the link is visible again

    TIMED OUT: success

    fail

    undefined

      686 |
      687 |   if (hardError) {
    > 688 |     throw new Error('TIMED OUT: ' + regex + '\n\n' + content + '\n\n' + lastErr)
          |           ^
      689 |   }
      690 |   return false
      691 | }

      at check (lib/next-test-utils.ts:688:11)
      at Object.<anonymous> (e2e/app-dir/app-client-cache/client-cache.test.ts:156:9)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Pull request closed without merging**